src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/Internal/HttpClientLogger.cs

```diff
@@ -0,0 +1,234 @@
+
+    private readonly ObjectPool<LogRecord> _logRecordPool =
+        PoolFactory.CreatePool(new LogRecordPooledObjectPolicy());
+
+    private readonly bool _logRequestStart;
+    private readonly bool _logResponseHeaders;
+    private readonly bool _logRequestHeaders;
+    private readonly bool _pathParametersRedactionSkipped;
+    private ILogger<HttpClientLogger> _logger;
+    private IHttpRequestReader _httpRequestReader;
+    private IHttpClientLogEnricher[] _enrichers;
+
+    public HttpClientLogger(
+        ILogger<HttpClientLogger> logger,
+        IHttpRequestReader httpRequestReader,
+        IEnumerable<IHttpClientLogEnricher> enrichers,
+        IOptions<LoggingOptions> options)
+    {
+        _logger = logger;
+        _httpRequestReader = httpRequestReader;
+        _enrichers = enrichers.ToArray();
+        var optionsValue = Throw.IfMemberNull(options, options.Value);
+
+        _logRequestStart = optionsValue.LogRequestStart;
+        _logResponseHeaders = optionsValue.ResponseHeadersDataClasses.Count > 0;
+        _logRequestHeaders = optionsValue.RequestHeadersDataClasses.Count > 0;
+        _pathParametersRedactionSkipped = optionsValue.RequestPathParameterRedactionMode == HttpRouteParameterRedactionMode.None;
+    }
+
+    [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "The logger shouldn't throw")]
+    public async ValueTask<object?> LogRequestStartAsync(HttpRequestMessage request, CancellationToken cancellationToken = default)
+    {
+        var logRecord = _logRecordPool.Get();
+
+        List<KeyValuePair<string, string>>? requestHeadersBuffer = null;
+        if (_logRequestHeaders)
+        {
+            requestHeadersBuffer = _headersPool.Get();
+        }
+
+        try
+        {
+            await _httpRequestReader.ReadRequestAsync(logRecord, request, requestHeadersBuffer, cancellationToken).ConfigureAwait(false);
+
+            if (_logRequestStart)
+            {
+                Log.OutgoingRequest(_logger, LogLevel.Information, logRecord);
+            }
+
+            return logRecord;
+        }
+        catch (Exception ex)
+        {
+            // If redaction is skipped, we can log unredacted request path; otherwise use "logRecord.Path" (even though it might not be set):
+            var pathToLog = _pathParametersRedactionSkipped
+                ? request.RequestUri?.AbsolutePath
+                : logRecord.Path;
+
+            Log.RequestReadError(_logger, ex, request.Method, request.RequestUri?.Host, pathToLog ?? string.Empty);
+
+            // Return back pooled objects (since the logRecord wasn't fully prepared):
+            _logRecordPool.Return(logRecord);
+
+            if (requestHeadersBuffer is not null)
+            {
+                _headersPool.Return(requestHeadersBuffer);
+            }
+
+            // Recommendation is to swallow the exception (logger shouldn't throw), so we don't re-throw here:
+            return null;
```

Comment:
As @CarnaViire pointed, a logger shouldn't re-throw (but it will change our current behavior)